evap/staff/importers/json.py

```diff
@@ -0,0 +1,448 @@
+    obj: str
+    message: str
+
+
+@dataclass
+class ImportStatistics:
+    name_changes: list[NameChange] = field(default_factory=list)
+    new_courses: list[Course] = field(default_factory=list)
+    new_evaluations: list[Evaluation] = field(default_factory=list)
+    updated_courses: list[Course] = field(default_factory=list)
+    updated_evaluations: list[Evaluation] = field(default_factory=list)
+    attempted_changes: list[Evaluation] = field(default_factory=list)
+    warnings: list[WarningMessage] = field(default_factory=list)
+
+    @staticmethod
+    def _make_heading(heading: str, separator: str = "-") -> str:
+        return f"{heading}\n{separator * len(heading)}\n"
+
+    @staticmethod
+    def _make_total(total: int) -> str:
+        return f"({total} in total)\n\n"
+
+    @staticmethod
+    def _make_stats(heading: str, objects: list) -> str:
+        log = ImportStatistics._make_heading(heading)
+        log += ImportStatistics._make_total(len(objects))
+        for obj in objects:
+            log += f"- {obj}\n"
+        log += "\n"
+        return log
+
+    def get_log(self) -> str:
+        log = self._make_heading("JSON IMPORTER REPORT", "=")
+        log += "\n"
+        log += f"Import finished at {now()}\n\n"
+
+        log += self._make_heading("Name Changes")
+        log += self._make_total(len(self.name_changes))
+        for name_change in self.name_changes:
+            log += f"- {name_change.old_first_name_given} {name_change.old_last_name} → {name_change.new_first_name_given} {name_change.new_last_name} (email: {name_change.email})\n"
+
+        log += self._make_stats("New Courses", self.new_courses)
+        log += self._make_stats("New Evaluations", self.new_evaluations)
+        log += self._make_stats("Updated Courses", self.updated_courses)
+        log += self._make_stats("Updated Evaluations", self.updated_evaluations)
+        log += self._make_stats("Attempted Changes", self.attempted_changes)
+
+        log += self._make_heading("Warnings")
+        log += self._make_total(len(self.warnings))
+        for warning in self.warnings:
+            log += f"- {warning.obj}: {warning.message}\n"
+
+        return log
+
+    def send_mail(self):
+        subject = "[EvaP] JSON importer log"
+
+        managers = UserProfile.objects.filter(groups__name="Manager", email__isnull=False)
+        if not managers:
+            return
+        mail = EmailMultiAlternatives(
+            subject,
+            self.get_log(),
+            settings.SERVER_EMAIL,
+            [manager.email for manager in managers],
+        )
+        mail.send()
+
+
+class JSONImporter:
+    DATETIME_FORMAT = "%d.%m.%Y %H:%M:%S"
+
+    def __init__(self, semester: Semester, default_course_end: str) -> None:
+        self.semester = semester
+        self.default_course_end = default_course_end
+        self.user_profile_map: dict[str, UserProfile] = {}
+        self.course_type_cache: dict[str, CourseType] = {
```

Comment:
I think we can improve the names here, how about `students_by_gguid` and `course_types_by_name` (and similar for the other variables)?